### DISCUSSION TITLE
Is it possible to add metadata to flags definitions?

### DISCUSSION THREAD

**Discussion by apodgorbunschih** [open-feature/flagd#1429] [Q&A]
Hello people 👋 
I would like to have some metadata returned with `flag` evaluation that is described in OpenFeature Specs: https://openfeature.dev/specification/types/#evaluation-details. I would like to return in the `metadata` things like "version" of the flag, "team" that is owning this flag. This data I will use for observability purposes on the JAVA/PHP SDK on the client side.

From the flagd implementation I see that the only `metadata` returned as the result of flag evaluation is the `scope`, and this is not set per `flag` but per "source".

ex: of the flags.json
```
{
  "$schema": "https://flagd.dev/schema/v0/flags.json",
  "flags": {
    "test-flag": {
      "defaultVariant": "true",
      "state": "ENABLED",
      "variants": {
        "false": false,
        "true": true
      },
      "metadata": {
        "version": "1.0.1",
        "team": "awesome-team"
      }
    }
  }
}
```

in the flagd code I would like something like that 
```
package model

import "encoding/json"

type Flag struct {
	State          string          `json:"state"`
	DefaultVariant string          `json:"defaultVariant"`
	Variants       map[string]any  `json:"variants"`
	Metadata       map[string]any  `json:"metadata"`
	Targeting      json.RawMessage `json:"targeting,omitempty"`
	Source         string          `json:"source"`
	Selector       string          `json:"selector"`
}
```

core/pkg/evaluator/json.go:312
```
func (je *Resolver) evaluateVariant(ctx context.Context, reqID string, flagKey string, evalCtx map[string]any) (
	variant string, variants map[string]interface{}, reason string, metadata map[string]interface{}, err error,
) {

...
	for key, value := range flag.Metadata {
		if key == SelectorMetadataKey {
			continue
		}
		metadata[key] = value
	}
...	
}
```

_If something like that make sense, I could try to contribute. Thanks in advance_ 🙇

**Comment by beeme1mr** [comment]
Hey @apodgorbunschih, sorry for the delayed response! I guess the notification from GitHub got buried in my email.

I was actually working on something related to this today. OpenFeature is working with OpenTelemetry on defining the semantics for [feature flag observiblity](https://github.com/open-feature/.github/issues/65). A draft PR open includes new metadata like a flag set ID, context ID, etc. For OpenFeature to adhere to this change, we'll need to access this information, and providers like flagd must be able to send it.

I believe the flagd flag configuration will likely need to be extended to include metadata as a top-level option, and for each flag. An example may look like this:

```
{
  "$schema": "https://flagd.dev/schema/v0/flags.json",
  "flags": {
    "test-flag": {
      "defaultVariant": "true",
      "state": "ENABLED",
      "variants": {
        "false": false,
        "true": true
      },
      "metadata": {
        "team": "awesome-team
      }
    }
  },
  "metadata": {
    "version": "1.0.1",
  }
}
```

Within the provider, the top level metadata could be stored as provider metadata and the flag level metadata could be returned as flag metadata. The reason I think it's important to use provider metadata is to reduce duplication and because provider metadata is available to all stages of a hook, where as flag metadata is only available in the after hook.

Do you think this would work for your use case?

**Reply by apodgorbunschih** [reply]
Hey @beeme1mr 👋 No worries at all, thanks a lot for the reply.

That is amazing that there are plans to improve observability, and this is exactly what we are looking for. Indeed, the flagd flag configuration will need to be adjusted to allow the metadata property. The schema that you presented will fulfill all our requirements.

@beeme1mr Do you have some date in mind when this could be available to the end user ? And is there something that we could help with ?

**Reply by beeme1mr** [reply]
I'm working on the requirements right now. I'll try and have the GitHub issues created by the end of the week. It's a relatively simple change, but the coordination across the project will need to be handled with care.

Here's a high-level overview of what needs to be done:

- Update the OpenFeature spec to expand provider metadata
- Update protobuf schemas to support provider/feature flag metadata
- Update JSON schema to support provider/feature flag metadata
- Update SDKs to support provider metadata
- Update flagd Providers to support provider/feature flag metadata
- Update OpenFeature spec with an OTel hook appendix
- Update telemetry hooks to consume the new data.

In the meantime, please take a look at these PRs. They help set the groundwork for this feature.

- https://github.com/open-telemetry/semantic-conventions/pull/1440
- https://github.com/open-feature/flagd/pull/1432

**Reply by apodgorbunschih** [reply]
Awesome, thank you for the provided details 🙇

**Reply by beeme1mr** [reply]
Hey @apodgorbunschih, I've opened a draft PR to add metadata to the flagd schema. It's marked as a draft to get feedback as I work on creating an implementation strategy. Hopefully, I'll be able to knock that out next week.

https://github.com/open-feature/flagd-schemas/pull/173

**Comment by apodgorbunschih** [comment]
Good morning @beeme1mr 
I see that the changes where accepted, implemented and release in [v0.11.6](https://github.com/open-feature/flagd/releases/tag/flagd%2Fv0.11.6). Big shoutout to you and the team for implementing that ❤️ 
Thanks 🙇
